### PR TITLE
Http client initial implementation

### PR DESCRIPTION
**Pull request by tobal**

### CODE REVIEW

File: lib/tests/test_client.py [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+SECRETS_PATH = '/kv/secret'
+ROLE_ID = '9eef5cf1-19d3-0cfb-b9bb-8b6248a17ece'
+SECRET_ID = '570bb5c9-2083-242d-5ba0-71224c1d4c48'
+CLIENT_TOKEN = 's.menL5xntpcw7RY2fMGDVU4Bo'
+VAULT_TOKEN = 's.AjtEPrapWWagEqXFtoNPOpaf'
+
+ROLE_ID_ENDPOINT = URL + '/v1/auth/approle/role/' + ROLE + '/role-id'
+SECRET_ID_ENDPOINT = URL + '/v1/auth/approle/role/' + ROLE + '/secret-id'
+LOGIN_ENDPOINT = URL + '/v1/auth/approle/login'
+PASSWORDS_ENDPOINT = URL + '/v1' + SECRETS_PATH
```
Comment (tobal) [inline lib/tests/test_client.py]: Will rename PASSWORDS_ENDPOINT to SECRETS_ENDPOINT

File: lib/client.py [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+                                                        data={'role-id': role_id, 'secret-id': secret_id})
+        secret = self._extract_data_from_endpoint(endpoint='/v1' + self._secrets_path,
+                                                  data_path='data.' + key,
+                                                  token=client_token,
+                                                  method='get')
+        return secret
+
+    def _extract_data_from_endpoint(self, endpoint, data_path, token, method, data=None):
+        headers = {'X-Vault-Token': token}
+        response = self._http_client.get(self._url + endpoint, headers=headers) if method == 'get' \
```
Comment (krajorama) [inline lib/client.py]: + logger.debug about what we're doing, but without the headers, cause that's sensitive info

File: lib/client.py [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    def __init__(self, vault_url, vault_token, secrets_path, auth_method, auth_params):
+        self._vault_url = vault_url
+        self._vault_token = vault_token
+        self._secrets_path = secrets_path
+        self._auth_method = auth_method
+        self._auth_params = auth_params
+
+    def instantiate(self):
+        if self._auth_method == 'app-role':
+            return AppRoleClient(api_params={'url': self._vault_url,
```
Comment (krajorama) [inline lib/client.py]: I'd rather the api_params was a namedtuple

File: lib/client.py [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+                                                  token=client_token,
+                                                  method='get')
+        return secret
+
+    def _extract_data_from_endpoint(self, endpoint, data_path, token, method, data=None):
+        headers = {'X-Vault-Token': token}
+        response = self._http_client.get(self._url + endpoint, headers=headers) if method == 'get' \
+            else self._http_client.post(self._url + endpoint, headers=headers, data=data)
+        if response.ok:
+            return reduce(dict.get, data_path.split('.'), json.loads(response.text))
```
Comment (krajorama) [inline lib/client.py]: cool!

Comment (krajorama) [inline lib/client.py]: + logger.debug just to indicate that we got a correct response

File: lib/client.py [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+    def from_config(cls, config):
+        return ClientFactory(
+            vault_url='http://{}:{}'.format(
+                config.get('vault', 'address'),
+                config.get('vault', 'port'),
+            ),
+            vault_token=config.get('vault', 'token'),
+            secrets_path=config.get('vault', 'secrets_path'),
+            auth_method=config.get('auth', 'method'),
+            auth_params=dict(role=config.get('auth', 'role'))
```
Comment (krajorama) [inline lib/client.py]: why do we need the dict?

Comment (tobal) [reply lib/client.py]: Spoke about it IRL, will make a separate "AppRoleAuthenticator" class

File: lib/tests/test_client.py [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+            },
+            'wrap_info': None,
+            'warnings': None,
+            'auth': None
+        }
+    ), True),
+}
+
+
+class StubHttpClient(object):
```
Comment (tobal) [inline lib/tests/test_client.py]: Talked about it in IRL: use MagicMock instead of this class

File: lib/client.py [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+            return AppRoleClient(api_params={'url': self._vault_url,
+                                             'secrets_path': self._secrets_path,
+                                             'vault_token': self._vault_token},
+                                 role=self._auth_params.get('role'),
+                                 http_client=requests)
+        else:
+            raise VaultException('Auth method not valid: {}'.format(self._auth_method))
+
+    @classmethod
+    def from_config(cls, config):
```
Comment (tobal) [inline lib/client.py]: Note to self: have a test for this function

Comment (tobal) [reply lib/client.py]: Done

File: lib/client.py [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+        return ClientFactory(
+            vault_url='http://{}:{}'.format(
+                config.get('vault', 'address'),
+                config.get('vault', 'port'),
+            ),
+            vault_token=config.get('vault', 'token'),
+            secrets_path=config.get('vault', 'secrets_path'),
+            auth_method=config.get('auth', 'method'),
```
[the remark this answers is not in the document]
Comment (krajorama) [reply lib/client.py]: IRL discussion: have a hashicorp_vault_token_authentication section. The pattern is : [<product>_<method>_authentication]. Which means safeguard should have the [safeguard_password_authentication] section.

File: lib/tests/test_client.py (hunk 6 above)
Comment (tobal) [reply lib/tests/test_client.py]: Done

File: lib/client.py (hunk 3 above)
Comment (tobal) [reply lib/client.py]: Done

File: lib/client.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (tobal) [reply lib/client.py]: Done

File: lib/client.py (hunk 4 above)
Comment (tobal) [reply lib/client.py]: Done

File: lib/client.py (hunk 2 above)
Comment (tobal) [reply lib/client.py]: Done

File: lib/client.py (hunk 5 above)
Comment (tobal) [reply lib/client.py]: Done

File: Pipfile [hunk 9]
```diff
@@ -0,0 +1,7 @@
+[packages]
+requests = ">=2.22.0"
```
[the remark this answers is not in the document]
Comment (tobal) [reply Pipfile]: Done

[the remark this answers is not in the document]
Comment (krajorama) [reply Pipfile]: I meant that the requirement can be removed altogether, to avoid packing it in the plugin ZIP.

File: Pipfile [hunk 10]
```diff
@@ -0,0 +1,7 @@
+[packages]
+requests = ">=2.18.4"
```
Comment (krajorama) [inline Pipfile]: please remove and update Pipfile.lock as well.

File: lib/tests/test_client.py [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+        client_factory,
+        vault_url=URL,
+        vault_token=VAULT_TOKEN,
+        secrets_path=SECRETS_PATH,
+        auth_method=AUTH_METHOD,
+        auth_params={'role': ROLE}
+    )
+
+
+@patch('requests.post', side_effect=POST_RESPONSES)
```
Comment (krajorama) [inline lib/tests/test_client.py]: cool!

File: lib/client.py [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+
+
+class Client(object):
+
+    def __init__(self, api_params, authenticator):
+        self._vault_url = api_params.vault_url
+        self._vault_token = api_params.vault_token
+        self._secrets_path = api_params.secrets_path
+        self._authenticator = authenticator
+        logger.debug('Http client initialized')
```
Comment (krajorama) [inline lib/client.py]: add vault_url for info

File: Pipfile (hunk 10 above)
Comment (tobal) [reply Pipfile]: Done

File: lib/client.py [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+                authenticator = AppRoleAuthenticator(api_params, role)
+            else:
+                raise VaultException('Missing "role" parameter for auth method {}'.format(self._auth_method))
+        else:
+            raise VaultException('Auth method not valid: {}'.format(self._auth_method))
+        return Client(api_params, authenticator)
+
+    @classmethod
+    def from_config(cls, config):
+        auth_method = config.get('hashicorp_vault', 'auth_method')
```
[the remark this answers is not in the document]
Comment (tobal) [reply lib/client.py]: Done

File: lib/client.py (hunk 12 above)
Comment (tobal) [reply lib/client.py]: Done

File: lib/client.py [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+        if auth_method == AuthMethods.APP_ROLE.value:
+            auth_params = dict(role=config.get('hashicorp_vault_approle_authentication', 'role'))
+        else:
+            raise VaultException('Auth method not valid: {}'.format(auth_method))
+        return ClientFactory(
+            vault_url='http://{}:{}'.format(
+                config.get('hashicorp_vault', 'address'),
+                config.get('hashicorp_vault', 'port'),
```
[the remark this answers is not in the document]
Comment (tobal) [reply lib/client.py]: Good point, done

**Pull request merged**